README.md
# xlreg_c

The C client for the xlReg cluster registry.

**xlreg_c**
makes it easy to create new small clusters some or all of whose members are
written in the programming language C.
xlreg_c is expected to be used primarily for testing, where
the ability to create and launch new, unique, distinct clusters in seconds
will be a major advantage.

In such a cluster, members use the **xlReg** public server to register and
exchange member IP addresses and other related information.
When a member finishes booting, it has
learned the cluster ID and the
IP addresses of its peers, their NodeIDs, and the public
RSA keys used to set up
encrypted communications
between members of the cluster and between cluster members and clients.

## The xlReg Service

<img src="img/xl-registration.jpg" alt="xl-registration" style="float:left" title="members registering with xlReg">

[xlReg](http://jddixon.github.io/xlReg_go)
is a service which provides its clients with unique nodeIDs and allows them
to register and join new clusters and collect configuration information
about existing clusters.

An **xlReg cluster** is a number of machines
which cooperate as a set of intercommunicating servers.  Each
such server has two RSA keys.  One, the **sig** key, is used for creating
digital signatures.  The other, the **comms** key, is used for encrypting
communications.  Most often the comms key is used for exchanging keys for
block ciphers. That is, the comms key is used only for agreeing on the
much faster block cipher key used during the rest of the communications
session.

Each xlReg cluster member has at least one IP address used for
intra-cluster communications, for communications between members.  In some
clusters each member will also have a second IP address used for
communications with cluster clients.

The xlReg server itself, its clients, and so the cluster members all are
or behave like
[XLattice nodes](http://jddixon.github.io/xlNode_go).
That is, they are self-contained fault-tolerant entities with some
persistent store.  These nodes may be located anywhere on the global
Internet, or on private networks with access to the network in which
an xlReg server resides.`

Like any XLattice node, the xlReg client can either be configured with some
specific keys at start-up or it will generate a new pair of keys and save
them to persistent store.

`## Registration

<img src="img/simple-cluster.jpg" alt="simple-cluster" style="float:right" title="small cluster, no clients">

A booting xlReg client establishes an encrypted connection to an xlReg server
(by default this is xlreg.xlattice.net).
If this is the client's first conversation with the xlReg server, the client
provides the server with configuration information including its RSA public
keys.  (The client does **not** register its secret RSA **private** keys.)
The configuration information is signed with the applicant's private
*sig" key.  The server uses the
client's RSA public key, provided as part of the configuration information, to
verify the client's digital signature on the configuration data.  This proves
that the configuration information is intact and that the applicant has the
private key corresponding to the public key.  In other words, the server has
in a certain sense confirmed the applicant's identity.  Having done this,
the xlReg server issues the client with a unique node ID, a 20-byte value,
completing client registration.

Once an xlReg registry client has identified itself to the registry it can

1. register new clusters or
2. join existing clusters or
3. get information on existing clusters.

The third option is used, for example, by nodes intending to become clients
of an existing cluster.

## Acquiring Configuration Data

After a new member has joined a cluster the server provides it
with cluster configuration data.  This includes

* the cluster name,
* the cluster ID, and
* the cluster size, the  maximum number of cluster members

It also includes for each current cluster member the member's

* nodeID,
* IP address(es),
* **sig** RSA public key, and
* **comms** RSA public key

<img src="img/cluster-with-clients.jpg" alt="cluster-with-clients" style="float:left" title="cluster with clients">

The new cluster member can then use this information to communicate with
other members.  Cluster members need no further assistance from the
xlReg server.

Cluster members will communicate with one another and clients
using protocols agreed-upon among themselves, but the xlReg registry
has no knowledge of such protocols.

**xlreg_c** is currently in development.  This is not
necessarily a description of running code; it is where we are going.

xlReg is an [XLattice](http://jddixon.github.io/xlattice_go/) project.  More
detailed (but somewhat dated) conceptual information on XLattice
is available at the [2003-2006 Java XLattice website](http://www.xlattice.org).

xlReg servers and clients communicate using the
[Gooble Protocol Buffers](http://code.google.com/p/protobuf/)-based
[xlReg protocol](http://jddixon.github.io/xlReg_go/xlReg_protocol.html).
An xlReg server has been in test for several months.

## Public Development Server

There is currently an xlReg server running on xlreg.xlattice.org:56789,
with the public keys shown in the registry credentials below.

    regCred {
        Name: xlReg
        ID: f2384769af97880208e76d012d8cf5b1b8be0b7b
        CommsPubKey: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDH6nJcf4cAbb1RWVvNS6YHjatZJQCerhmISdyhBDmNJAUuBpr+LcrOe0d+kxEPQB+IbNKfvj7ROd8Aet74s5sJ0AcZyFXLFPvt6eugLCzrXmgDo60qoUtZgxaxisQvsoAZB2FR+nqJAG8E84o3rGD1bNM/2WgjvuWoLpfy+RE8hCUFku7yo9JdOunhc4rpCkBeX9M7g8oWPanHDWMrSwsE8CBf/4pXlEiNNd3YHkwQseUrkCvpLCDuYYV8xjzAYexoBWrVGwpeTO/QtwdebwPxa7ArEQfPwgA5hn6S9iy9+EK6zay7OKqKioaOoYDlJ+Z1rAIi6dBJXdRVHGsMc/kl

        SigPubKey: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCk84/cQ/YnxOLElukZpeQ9QT9FdgcYujmIUGOaFcWtxUJJoSXgBterjgJ+fE5vvNN/kdX97Ps9Mz8PHTAUyClUn5DMxFYApDp/UwnBk0MNqaCnqt3p9ymflIZiex1+/pSWcSpqZYW9hIUQddEGWG6a4dWyD2lhoD1II5Iuj0xMgN167fgsA7otgd/wTitp2jrJE+0cZ4eQShpkuMd9BvNJmkMoyF5PzBnMcRxZ9RvM4zEOe23GRSGGAuPUvTaVEeYY3TVFgF4YfXiYOrPQynS6B7+fnWXu2XDQrwP3mZxYAYJqL5nZvQypCilc8dDxFunf5B+GxdAwqyQbTNwM3ksH

        EndPoints {
             TcpEndPoint: 54.186.197.123:56789
        }
        Version: 1.9.20
    }

The IP address in the reg credentials above is of course that of xlreg.xlattice.org.

## Server Implementation

The xlReg server and its supporting libraries are written in
[the Go programming language](http://golang.org), a high-performance
systems programming language incorporating certain features of
[Tony Hoare](http://en.wikipedia.org/wiki/Tony_Hoare)'s
revolutionary **CSP**,
[Communicating Sequential Processes](http://www.usingcsp.com).

Google's Go language is a stable product.  It has been available for
several years and the compiler is widely acknowledged to be very fast
and to generate fast, compact code.  The Go language itself --
the compiler and its libraries -- is free and open source.

## Server Source Code

Software developers can currently obtain the xlReg server source code at
[xlReg_go's Github project page](https://github.com/jddixon/xlReg_go/)
so that if you are impatient to begin using xlReg you can build and operate
your own server.

## Client Source Code

This is best obtained through
[this client's Github project page](https://github.com/jddixon/xlreg_c/).

## Project Status

The xlReg server is currently a good beta: it is reasonably reliable and
there are no known serious bugs.

**This client is itself pre-alpha:** some minimal code exists, but nothing
beyond reading the registry credentials (`regCred.dat`) and going through
the initial hello/reply session with the server, xlReg.
## On-line Documentation

More information on the **xlreg_c** project can be found
[here](https://jddixon.github.io/xlreg_c)


CHANGES
# xlreg_c/CHANGES

v0.1.5
    2016-11-13
        * update docs for xlReg 1.9.20                              SLOC 1023
v0.1.4
    2016-09-24
        * standardize licensing                                     SLOC 1021
v0.1.3
    2016-07-11
        * add ghpDoc/{css,img}                                      SLOC 1002
v0.1.2
    2016-06-18
        * drop autotools, add waf; add .dvcz, bkp2U, other bash     SLOC  494
v0.1.1
    2016-03-11
        * options working                                           SLOC  263
v0.1.0
    2016-03-06
        * strip out autotools, add waf                              SLOC  264
v0.0.7
    2016-02-29
        * add countem, update LICENSE                               SLOC  280
v0.0.6
    2015-12-21
        * fixed dead link(s)
v0.0.5
    2015-12-11
        * working -V/--stepVersion options                          SLOC  200
v0.0.4
    2015-11-30
        * minor updates to *.md                                     SLOC  146
v0.0.3
    2015-10-10
        * update {README,index}.md for xlReg 1.9.10:
v0.0.2
    2015-08-01
        * update README/index.md for xlReg v1.9.9                   SLOC  121
v0.0.1
    2015-06-16
        * update README/index.md, .gitignore                        SLOC  121














include/xlreg_errors.h
/* xlreg_c/src/xlreg_errors.h */

#ifndef XLREG_ERRORS_H
#define XLREG_ERRORS_H

#define BAD_ATTRS_LINE = -1
#define BAD_VERSION = -2
#define CANT_FIND_CLUSTER_BY_ID = -3
#define CANT_FIND_CLUSTER_BY_NAME = -4
#define CLUSTER_MEMBER_ID_IN_USE = -5
#define CLUSTER_MEMBERS_MUST_HAVE_END_POINT = -6
#define CLUSTER_MEMBER_NAME_IN_USE = -7
#define CLUSTER_MUST_HAVE_MEMBER = -8
#define CLUSTER_MUST_HAVE_TWO = -9
#define ID_ALREADY_IN_USE = -10
#define ILL_FORMED_CLUSTER = -11
#define ILL_FORMED_CLUSTER_MEMBER = -12
#define ILL_FORMED_REG_CRED = -13
#define MEMBER_MUST_HAVE_END_POINT = -14
#define MISSING_CLOSING_BRACE = -15
#define MISSING_CLUSTER_NAME_OR_ID = -16
#define MISSING_END_POINTS_SECTION = -17
#define MISSING_MEMBERS_LIST = -18
#define MISSING_NODE = -19
#define MISSING_PRIVATE_KEY = -20
#define MISSING_REG_NODE_LINE = -21
#define MISSING_SERVER_INFO = -22
#define NAME_ALREADY_IN_USE = -23
#define NIL_CLUSTER = -24
#define NIL_CLUSTER_MEMBER = -25
#define NIL_NODE = -26
#define NIL_PRIVATE_KEY = -27
#define NIL_REGISTRY = -28
#define NIL_REG_NODE = -29
#define NIL_TOKEN = -30
#define RCVD_INVALID_MSG_FOR_STATE = -31
#define TAG_OUT_OF_RANGE = -32
#define UNEXPECTED_MSG_TYPE = -33
#define WRONG_NUMBER_OF_BYTES_IN_ATTRS = -34

char *XLREG_ERRORS[] = {
    "badly formed attrs line",
    "badly formated VERSION",
    "cannot find cluster with this ID",
    "cannot find cluster with this name",
    "cluster member id already in use",
    "cluster members must have at least one endPoint",
    "cluster member name already in use",
    "cluster must have a member!",
    "cluster must have at least two members",
    "ID already in use",
    "ill-formed cluster serialization",
    "ill-formed cluster member serialization",
    "ill-formed regCred serialization",
    "member must have at least one endPoint",
    "missing closing brace",
    "missing cluster name or ID",
    "missing endPoints section",
    "missing members list",
    "missing node parameter",
    "missing private key line",
    "missing regNode line",
    "missing server info",
    "name already in use",
    "nil cluster argument",
    "nil cluster member argument",
    "nil node argument",
    "nil private key argument",
    "nil registry argument",
    "nil RegNode argument",
    "nil XLRegMsg_Token argument",
    "invalid msg type for current state",
    "message tag out of range",
    "unexpected message type",
    "wrong number of bytes in attrs",
};

#endif
